[RFC-0010] Add workload identity support to Google Pub/Sub notifier

cappyzawa opened the pull request:
Part of https://github.com/fluxcd/flux2/issues/5022.

> notification-controller > Google Pub/Sub

matheuscscp commented on internal/notifier/google_helpers.go:
```diff
@@ -0,0 +1,88 @@
+
+const (
+	// GCP Pub/Sub scope for authentication
+	gcpPubSubScope = "https://www.googleapis.com/auth/pubsub"
```
We don't need to pass scopes in GCP, only Azure needs it. In the `auth/gcp` lib we pass a wildcard scope that works for all GCP APIs. Azure doesn't have the same wildcard scope :/

matheuscscp commented on internal/notifier/google_helpers.go:
```diff
@@ -0,0 +1,88 @@
+	ServiceAccountName string
+	ProviderName       string
+	ProviderNamespace  string
+	ProxyURL           string
```
Since we are introducing proxy support for authentication, we could also introduce proxy support for the whole operation, i.e. for interacting with the Pub/Sub API itself. For this we need to build a custom Google transport, like this:

https://github.com/fluxcd/source-controller/blob/c2b572bae03a53cef4355389494ea71097c6abfe/pkg/gcp/gcp.go#L90-L145

matheuscscp commented on docs/spec/v1beta3/providers.md:
```diff
@@ -778,6 +778,75 @@ stringData:
+  located in the `flux-system` namespace with name `notification-controller`.
+
+- Configure workload identity with notification-controller as described in the
+  docs [here](/flux/installation/configuration/workload-identity/).
```
```suggestion
  docs [here](/flux/integrations/gcp/).
```

matheuscscp replied:
Let's link also directly to https://fluxcd.io/flux/integrations/gcp/#for-google-cloud-pubsub, which is the specific section for Pub/Sub

cappyzawa commented on internal/notifier/google_pubsub.go:
```diff
@@ -70,14 +73,26 @@ func NewGooglePubSub(projectID, topicID, jsonCreds string, attrs map[string]stri
+	}
+
+	shouldUseWorkloadIdentity := len(jsonCreds) == 0 && identityConfig.IsValid()
+	if shouldUseWorkloadIdentity {
```
Since we already have a factory, I changed it so the factory creates the options and we just pass those along.Since we already have a factory, I changed it so the factory creates the options and we just pass those along.

[1d04f98](https://github.com/fluxcd/notification-controller/pull/1154/commits/1d04f989bad63e635ab1f97822007757eea58c3b)

cappyzawa commented on internal/notifier/google_helpers.go:
```diff
@@ -0,0 +1,106 @@
+
+	if opts.Token != "" {
+		clientOpts = append(clientOpts, option.WithCredentialsJSON([]byte(opts.Token)))
+	} else if opts.TokenClient != nil {
```
No. I removed this handling.

cappyzawa commented:
I’m looking into the code to try to find the cause.

matheuscscp commented:
The logs above are for object-level. The same happens at the controller level:

```json
{
    "level": "error",
    "ts": "2025-08-02T17:35:23.582Z",
    "logger": "event-server",
    "msg": "failed to send notification",
    "eventInvolvedObject": {
        "kind": "OCIRepository",
        "namespace": "flux-system",
        "name": "flux-system",
        "uid": "f2dde3e6-6411-498a-b26c-1c14c00b36ac",
        "apiVersion": "source.toolkit.fluxcd.io/v1",
        "resourceVersion": "109404"
    },
    "alert": {
        "name": "pubsub",
        "namespace": "flux-system",
        "providerName": "pubsub"
    },
    "error": "error publishing event to topic projects/flux-gitops-playground/topics/test-flux: rpc error: code = Unauthenticated desc = transport: per-RPC creds failed due to error: failed to create provider access token for the controller: oauth2/google: invalid response from Secure Token Server: Post \"https://sts.googleapis.com/v1/token\": context canceled"
}
```

Gives strength to the theory of a bad context being passed somewhere

matheuscscp commented:
It's not here, I added a log and it doesn't get printed:

```go
		if ctx.Err() != nil {
			log.FromContext(ctx).Error(ctx.Err(), "context error while building GCP client options")
		}
		tokenSource := gcp.NewTokenSource(ctx, authOpts...)
		clientOpts = append(clientOpts, option.WithTokenSource(tokenSource))
```

cappyzawa commented:
### Problem

The workload identity implementation was causing "context canceled" errors in matheus's test environment:

```
"error": "rpc error: code = Unauthenticated desc = transport: per-RPC creds failed due to error: failed to get service account 'flux-system/pubsub': Timeout: failed waiting for *v1.ServiceAccount Informer to sync"

"error": "rpc error: code = Unauthenticated desc = transport: per-RPC creds failed due to error: failed to create kubernetes token for service account 'flux-system/pubsub': context canceled"
```

### Suspected Cause

The issue was likely caused by a context timeout mismatch in the notification pipeline:

1. **Event Server Handler**: Creates a 15-second timeout context for HTTP request processing
2. **Notifier Creation**: The context flows through `dispatchNotification` → `getNotificationParams` → `createNotifier` → `buildGCPClientOptions`
3. **TokenSource Storage**: `gcp.NewTokenSource(ctx, ...)` stores the 15-second context in its struct (Go anti-pattern)
4. **Workload Identity Setup**: When `TokenSource.Token()` is called later, it uses the stored context for:
   - ServiceAccount informer synchronization
   - Kubernetes token creation via `client.SubResource("token").Create(ctx, ...)`
5. **Timeout**: These operations can take longer than 15 seconds, causing context cancellation

### Context Flow Analysis (generated by AI)

```
HTTP Request → handleEvent (15s timeout)
             ↓
             dispatchNotification
             ↓
             getNotificationParams  
             ↓
             createNotifier
             ↓
             buildGCPClientOptions → gcp.NewTokenSource(15s_ctx)
                                                        ↓
                                     [context stored in struct]
                                                        ↓
             [goroutine starts with fresh context]
             ↓
             notifier.Post(fresh_ctx) → TokenSource.Token()
                                                    ↓
                                        [uses old 15s context]
                                                    ↓
                                        auth.GetAccessToken(old_ctx)
                                                    ↓
                                        ServiceAccount operations → TIMEOUT!
```

### Solution

Modified `buildGCPClientOptions` to use `context.Background()` instead of the HTTP request context:

1. **Function Signature**: Changed to `buildGCPClientOptions(ctx context.Context, opts notifierOptions)` for explicit context control
2. **Background Context**: Call with `context.Background()` to avoid timeout during workload identity setup
3. **Safety**: The actual notification sending timeout is controlled separately in the goroutine with a fresh context

cappyzawa commented:
### Additional Context

For context, this issue highlights a common Go design pattern consideration. The [official Go documentation](https://pkg.go.dev/context) recommends:

> Do not store Contexts inside a struct type; instead, pass a Context explicitly to each function that needs it.

This is discussed further in the [Go blog post about context and structs](https://go.dev/blog/context-and-structs). The `gcp.NewTokenSource()` function stores the context in its struct to match the `oauth2.TokenSource` interface, which predates Go's context patterns. This design pattern can lead to the timeout issues we encountered, where a short-lived context gets stored and used later when it may have already been canceled.

Our fix works around this by providing a long-lived background context specifically for the TokenSource, while keeping the request-scoped timeout control separate in the notification goroutine.

matheuscscp commented:
The `oauth2.TokenSource` interface observation is most correct :+1: We'd love to be able to pass a `ctx` in the right place here, but the old, wide-spread interface does not allow us :cry:

matheuscscp commented:
> Better yet, all of the code inside `buildGCPClientOptions` can actually run in `Post()`. The only code that needs to be placed inside `publish()` is code that we cannot run successfully in CI, which is `pubsub.NewClient()` and `client.Topic().Publish().Get()`. This way `Post()` is unit-testable through the `googlePubSubPostTestCase` mock.

Turns out that this is not true, and to achieve using the `ctx` created on the beginning of the goroutine quite a refactor is needed, see collapsed diff

<details>

```diff
diff --git a/internal/notifier/factory.go b/internal/notifier/factory.go
index 062ecae..f2dfdb8 100644
--- a/internal/notifier/factory.go
+++ b/internal/notifier/factory.go
@@ -263,21 +263,7 @@ func googleChatNotifierFunc(opts notifierOptions) (Interface, error) {
 }
 
 func googlePubSubNotifierFunc(opts notifierOptions) (Interface, error) {
-	// NOTE: Use background context for GCP client options to avoid workload identity timeout.
-	// opts.Context has a 15-second timeout which is too short for ServiceAccount informer
-	// sync and token creation. The actual notification timeout is handled separately.
-	clientOpts, err := buildGCPClientOptions(context.Background(), opts)
-	if err != nil {
-		return nil, err
-	}
-
-	return NewGooglePubSub(
-		opts.Context,
-		opts.URL,
-		opts.Channel,
-		opts.Headers,
-		clientOpts,
-	)
+	return NewGooglePubSub(&opts)
 }
 
 func webexNotifierFunc(opts notifierOptions) (Interface, error) {
diff --git a/internal/notifier/google_pubsub.go b/internal/notifier/google_pubsub.go
index 0bf92d0..52b2bb7 100644
--- a/internal/notifier/google_pubsub.go
+++ b/internal/notifier/google_pubsub.go
@@ -23,7 +23,6 @@ import (
 	"fmt"
 
 	"cloud.google.com/go/pubsub"
-	"google.golang.org/api/option"
 	kerrors "k8s.io/apimachinery/pkg/util/errors"
 	"sigs.k8s.io/controller-runtime/pkg/log"
 
@@ -33,18 +32,13 @@ import (
 type (
 	// GooglePubSub holds a Google Pub/Sub client and target topic.
 	GooglePubSub struct {
-		topicID   string
-		attrs     map[string]string
-		topicName string
-
 		client interface {
-			publish(ctx context.Context, topicID string, eventPayload []byte, attrs map[string]string) (serverID string, err error)
+			publish(ctx context.Context, eventPayload []byte) error
 		}
 	}
 
 	googlePubSubClient struct {
-		projectID string
-		opts      []option.ClientOption
+		opts *notifierOptions
 	}
 )
 
@@ -53,29 +47,15 @@ var _ Interface = &GooglePubSub{}
 
 // NewGooglePubSub creates a Google Pub/Sub client tied to a specific
 // project and topic using the provided client options.
-func NewGooglePubSub(ctx context.Context, projectID, topicID string, attrs map[string]string,
-	clientOpts []option.ClientOption) (*GooglePubSub, error) {
-	if projectID == "" {
+func NewGooglePubSub(opts *notifierOptions) (*GooglePubSub, error) {
+	if opts.URL == "" {
 		return nil, errors.New("GCP project ID cannot be empty")
 	}
-	if topicID == "" {
+	if opts.Channel == "" {
 		return nil, errors.New("GCP Pub/Sub topic ID cannot be empty")
 	}
-	if len(attrs) == 0 {
-		attrs = nil
-	}
-
-	client := &googlePubSubClient{
-		projectID: projectID,
-		opts:      clientOpts,
-	}
-
-	return &GooglePubSub{
-		topicID:   topicID,
-		attrs:     attrs,
-		topicName: fmt.Sprintf("projects/%s/topics/%s", projectID, topicID),
-		client:    client,
-	}, nil
+	client := &googlePubSubClient{opts}
+	return &GooglePubSub{client}, nil
 }
 
 // Post posts Flux events to a Google Pub/Sub topic.
@@ -90,24 +70,21 @@ func (g *GooglePubSub) Post(ctx context.Context, event eventv1.Event) error {
 		return fmt.Errorf("error json-marshaling event: %w", err)
 	}
 
-	serverID, err := g.client.publish(ctx, g.topicID, eventPayload, g.attrs)
-	if err != nil {
-		return fmt.Errorf("error publishing event to topic %s: %w", g.topicName, err)
-	}
-
-	// debug log
-	log.FromContext(ctx).V(1).Info("Event published to GCP Pub/Sub topic",
-		"topic", g.topicName,
-		"server message id", serverID)
-
-	return nil
+	return g.client.publish(ctx, eventPayload)
 }
 
-func (g *googlePubSubClient) publish(ctx context.Context, topicID string, eventPayload []byte, attrs map[string]string) (serverID string, err error) {
-	var client *pubsub.Client
-	client, err = pubsub.NewClient(ctx, g.projectID, g.opts...)
+func (g *googlePubSubClient) publish(ctx context.Context, eventPayload []byte) error {
+	projectID := g.opts.URL
+	topicID := g.opts.Channel
+
+	// Build client.
+	opts, err := buildGCPClientOptions(ctx, *g.opts)
+	if err != nil {
+		return err
+	}
+	client, err := pubsub.NewClient(ctx, projectID, opts...)
 	if err != nil {
-		return
+		return err
 	}
 	defer func() {
 		if closeErr := client.Close(); closeErr != nil {
@@ -118,12 +95,28 @@ func (g *googlePubSubClient) publish(ctx context.Context, topicID string, eventP
 			}
 		}
 	}()
-	serverID, err = client.
-		Topic(topicID).
+
+	// Publish the event to the topic.
+	attrs := g.opts.Headers
+	if len(attrs) == 0 {
+		attrs = nil
+	}
+	topic := fmt.Sprintf("projects/%s/topics/%s", projectID, topicID)
+	serverID, err := client.
+		Topic(topic).
 		Publish(ctx, &pubsub.Message{
 			Data:       eventPayload,
 			Attributes: attrs,
 		}).
 		Get(ctx)
-	return
+	if err != nil {
+		return fmt.Errorf("error publishing to GCP Pub/Sub topic %s: %w", topic, err)
+	}
+
+	// Emit debug log.
+	log.FromContext(ctx).V(1).Info("Event published to GCP Pub/Sub topic",
+		"topic", topic,
+		"server message id", serverID)
+
+	return nil
 }
diff --git a/internal/notifier/google_pubsub_test.go b/internal/notifier/google_pubsub_test.go
index a94da52..d9e491a 100644
--- a/internal/notifier/google_pubsub_test.go
+++ b/internal/notifier/google_pubsub_test.go
@@ -19,7 +19,6 @@ package notifier
 import (
 	"context"
 	"errors"
-	"fmt"
 	"testing"
 
 	. "github.com/onsi/gomega"
@@ -50,60 +49,18 @@ func TestNewGooglePubSub(t *testing.T) {
 			topicID:     "",
 			expectedErr: errors.New("GCP Pub/Sub topic ID cannot be empty"),
 		},
-		{
-			name:              "topic name is stored properly",
-			projectID:         "project-id",
-			topicID:           "topic-id",
-			expectedTopicName: "projects/project-id/topics/topic-id",
-		},
-		{
-			name:              "client options are stored properly",
-			projectID:         "project-id",
-			topicID:           "topic-id",
-			expectedTopicName: "projects/project-id/topics/topic-id",
-			clientOpts:        []option.ClientOption{option.WithCredentialsJSON([]byte("json credentials"))},
-		},
-		{
-			name:              "non-empty attributes are stored properly",
-			projectID:         "project-id",
-			topicID:           "topic-id",
-			expectedTopicName: "projects/project-id/topics/topic-id",
-			attrs:             map[string]string{"foo": "bar"},
-			expectedAttrs:     map[string]string{"foo": "bar"},
-		},
-		{
-			name:              "empty attributes are stored properly",
-			projectID:         "project-id",
-			topicID:           "topic-id",
-			expectedTopicName: "projects/project-id/topics/topic-id",
-			attrs:             map[string]string{},
-			expectedAttrs:     nil,
-		},
 	}
 
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
 			g := NewWithT(t)
-			ctx := context.Background()
-			provider, err := NewGooglePubSub(ctx, tt.projectID, tt.topicID, tt.attrs, tt.clientOpts)
+			provider, err := NewGooglePubSub(&notifierOptions{
+				URL:     tt.projectID,
+				Channel: tt.topicID,
+			})
 
-			if tt.expectedErr != nil {
-				g.Expect(err).To(Equal(tt.expectedErr))
-				g.Expect(provider).To(BeNil())
-			} else {
-				g.Expect(err).To(BeNil())
-				g.Expect(provider).NotTo(BeNil())
-
-				g.Expect(provider.topicID).To(Equal(tt.topicID))
-				g.Expect(provider.attrs).To(Equal(tt.expectedAttrs))
-				g.Expect(provider.topicName).To(Equal(tt.expectedTopicName))
-
-				g.Expect(provider.client).NotTo(BeNil())
-
-				client := provider.client.(*googlePubSubClient)
-				g.Expect(client).NotTo(BeNil())
-				g.Expect(client.opts).To(Equal(tt.clientOpts))
-			}
+			g.Expect(err).To(Equal(tt.expectedErr))
+			g.Expect(provider).To(BeNil())
 		})
 	}
 }
@@ -123,14 +80,11 @@ type googlePubSubPostTestCase struct {
 	g *WithT
 }
 
-func (tt *googlePubSubPostTestCase) publish(ctx context.Context, topicID string, eventPayload []byte, attrs map[string]string) (serverID string, err error) {
+func (tt *googlePubSubPostTestCase) publish(ctx context.Context, eventPayload []byte) error {
 	tt.g.THelper()
 	tt.publishExecuted = true
-	tt.g.Expect(topicID).To(Equal(tt.topicID))
 	tt.g.Expect(string(eventPayload)).To(Equal(tt.expectedEventPayload))
-	tt.g.Expect(attrs).To(Equal(tt.attrs))
-	// serverID is only used in a debug log for now, there's no way to assert it
-	return "", tt.publishErr
+	return tt.publishErr
 }
 
 func TestGooglePubSubPost(t *testing.T) {
@@ -151,11 +105,11 @@ func TestGooglePubSubPost(t *testing.T) {
 			publishShouldExecute: false,
 		},
 		{
-			name:                 "publish error is wrapped and relayed",
+			name:                 "publish error is relayed",
 			expectedEventPayload: `{"involvedObject":{},"severity":"","timestamp":null,"message":"","reason":"","reportingController":""}`,
 			topicName:            "projects/projectID/topics/topicID",
 			publishErr:           errors.New("publish error"),
-			expectedErr:          fmt.Errorf("error publishing event to topic projects/projectID/topics/topicID: %w", errors.New("publish error")),
+			expectedErr:          errors.New("publish error"),
 			publishShouldExecute: true,
 		},
 		{
@@ -173,10 +127,7 @@ func TestGooglePubSubPost(t *testing.T) {
 			tt.g = g
 
 			topic := &GooglePubSub{
-				client:    tt,
-				topicID:   tt.topicID,
-				attrs:     tt.attrs,
-				topicName: tt.topicName,
+				client: tt,
 			}
 
 			err := topic.Post(context.Background(), tt.event)
```

</details>

cappyzawa commented:
Sorry for making you do most of the implementation…! (good morning 🌞 )
I’ll go ahead and incorporate your fix—thanks!

matheuscscp commented:
> Sorry for making you do most of the implementation…! (good morning 🌞 ) I’ll go ahead and incorporate your fix—thanks!

There's nothing to be sorry about! It's my mistake not guiding you correctly beforehand, I definitely did not predict this issue we found out, it's for sure something to keep in mind more often. You are helping us out tons here! :grin:

cappyzawa commented:
@matheuscscp I’ve merged your changes in 9021d14.
Could you check if the diff looks as intended? :pray:

That was an awesome refactor—thanks a lot!

cappyzawa commented:
If you've finished reviewing the changes, let me know and I'll squash the commits 🙏

matheuscscp commented:
> see collapsed diff

I did a last-minute edit on this comment after catching a problem (passing wrong arg to `client.Topic()`), it looks like you fixed it for me, is that correct?

```diff
diff --git a/internal/notifier/google_pubsub.go b/internal/notifier/google_pubsub.go
index 52b2bb7..0dd4ef9 100644
--- a/internal/notifier/google_pubsub.go
+++ b/internal/notifier/google_pubsub.go
@@ -103,7 +103,7 @@ func (g *googlePubSubClient) publish(ctx context.Context, eventPayload []byte) e
 	}
 	topic := fmt.Sprintf("projects/%s/topics/%s", projectID, topicID)
 	serverID, err := client.
-		Topic(topic).
+		Topic(topicID).
 		Publish(ctx, &pubsub.Message{
 			Data:       eventPayload,
 			Attributes: attrs,
```

matheuscscp reviewed (approved):
LGTM! 🚀 

Thanks very much @cappyzawa ❤️ 

I was able to test both *controller-level* and *object-level* workload identity for the commit https://github.com/fluxcd/notification-controller/pull/1154/commits/9021d1458807f3e5a61defd43d733d58251b95aa:

```json
{
    "involvedObject": {
        "kind": "OCIRepository",
        "namespace": "flux-system",
        "name": "flux-system",
        "uid": "f2dde3e6-6411-498a-b26c-1c14c00b36ac",
        "apiVersion": "source.toolkit.fluxcd.io/v1",
        "resourceVersion": "247276"
    },
    "severity": "info",
    "timestamp": "2025-08-02T22:21:54Z",
    "message": "stored artifact with revision 'latest@sha256:<redacted>' from 'oci://<redacted>', origin source 'https://<redacted>', origin revision 'refs/heads/main@sha1:<redacted>'",
    "reason": "NewArtifact",
    "metadata": {
        "authMode": "controller-level",
        "cluster": "preview",
        "revision": "latest@sha256:<redacted>"
    },
    "reportingController": "source-controller",
    "reportingInstance": "source-controller-775f75d5b4-6rll5"
}
```

```json
{
    "involvedObject": {
        "kind": "OCIRepository",
        "namespace": "flux-system",
        "name": "flux-system",
        "uid": "f2dde3e6-6411-498a-b26c-1c14c00b36ac",
        "apiVersion": "source.toolkit.fluxcd.io/v1",
        "resourceVersion": "246335"
    },
    "severity": "info",
    "timestamp": "2025-08-02T22:18:58Z",
    "message": "stored artifact with revision 'latest@sha256:<redacted>' from 'oci://<redacted>', origin source 'https://<redacted>', origin revision 'refs/heads/main@sha1:<redacted>'",
    "reason": "NewArtifact",
    "metadata": {
        "authMode": "object-level",
        "cluster": "preview",
        "revision": "latest@sha256:<redacted>"
    },
    "reportingController": "source-controller",
    "reportingInstance": "source-controller-775f75d5b4-6rll5"
}
```

I added the `authMode` metadata key to the event through the `.spec.eventMetadata` API:

```yaml
apiVersion: notification.toolkit.fluxcd.io/v1beta3
kind: Alert
metadata:
  name: pubsub
  namespace: flux-system
spec:
  providerRef:
    name: pubsub
  eventMetadata:
    cluster: preview
    authMode: controller-level # and object-level while testing .spec.serviceAccountName set on the Provider object
  eventSeverity: info
  eventSources:
    - kind: OCIRepository
      name: '*'
```
